### ISSUE TITLE
Proposal: Re-implement in Go

### ISSUE THREAD

**Issue by unguiculus** [helm/chart-testing#29]
I'm tinkering with re-implementing chart-testing in Go. I've already thought about this when I wrote it in Bash and, looking at it in hindsight, I think I should have done it right from the start. I probably didn't expect we'd end up with a script containing more than 500 lines of code. Too much for a Bash script IMHO. On the other hand, Python is excellent for scripting, you can debug it, and IDE support is very good. It's easier to write and maintain than Bash.

The project is getting more popular with more PRs and issues coming in. However, adding new features to a Bash script can get messy. There are too many different styles and it's too easy to make mistakes, especially when you don't have a debugger. 

Configuration is currently not ideal. We have a mixture of CLI args and environment variables. Config and args parsing would be much easier in Python. Without any special effort, we could support CLI args, env vars, and config files.

Overall, we'd get a better understandable and maintainable code and probably lower the threshold for contributions.

Opinions?

cc @helm/charts-maintainers 

----

**Update:** The re-implementation will happen in Go!

**Comment by rimusz** [comment]
yes, totally makes sense 👍

**Comment by scottrigby** [comment]
Summarizing from from chat, the idea of possibly using Go. For reasons. But I don't have a strong opinion on this - IMO it's up to your preference @unguiculus 🤩

For the compile step, I reco [multistage Docker files](https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds). For distribution, would love to use [goreleaser](https://github.com/goreleaser/goreleaser) (that can also solve the question of [committing the vendor dir with dep](https://github.com/scottrigby/example-dep-no-vendor)).

Minor consideration: I could pitch in more quickly if this were Go than Python 🤷‍♂️😄

**Comment by rimusz** [comment]
++1 for go

**Comment by unguiculus** [comment]
Initially, I chatted with @mattfarina whether Python or Go should be used. We leaned towards Python. But the more I think about it the more I'd go with Go 😉. So, Go will be the winner.

@scottrigby Im sure we'll find something for you to pitch in. 😀

**Comment by prydonius** [comment]
In the charts chat, I came to the conclusion that Go might not be the best language for this, given it's going to be a lot of `exec.Command`s, it might be much more readable in something like Python. At the end of the day, as @mattfarina said, we should choose whatever language the people working on this will be more comfortable in.

**Comment by unguiculus** [comment]
@prydonius It doesn't make much of a difference whether you launch a process in Python or Go. The code is similar. Now that I've played with both Python and Go I'm convinced that Go would be the better option. And I'd like the fact that we'll get a single binary that can easily be run on various OSes. Also, I've already invested quite some time in the Go variant and don't want to go back to Python anymore at this point.

**Comment by rimusz** [comment]
Yes, let's go for `go` 😀

**Comment by jdolitsky** [comment]
Re: discussion in charts call - another alternative is to use the Robot acceptance testing framework: http://robotframework.org/

This allows you to define your tests in easy to understand, english test cases such as "Chart passes lint" - then this test can be implemented behind the scenes in a mini Python library that launches a subprocess.

The robot console output looks something like:
```
==============================================================================
Acceptance Tests.Required-Env-Vars :: Tests to verify required sets of env ...
==============================================================================
Crashes if missing CHART_REF or CHART_NAME                            | PASS |
------------------------------------------------------------------------------
Will not crash if has one of CHART_REF or CHART_NAME                  | PASS |
------------------------------------------------------------------------------
Will use CHART_REF over CHART_NAME if both are provided               | PASS |
```

It also spits out a nice HTML test report that looks like this:
<img width="1132" alt="screen shot 2018-10-02 at 11 55 39 am" src="https://user-images.githubusercontent.com/393494/46363984-2242b500-c63a-11e8-9a55-3d458f237947.png">

Here is example of ChartMuseum acceptance tests robot HTML report (feel free to click around): https://143-104032019-gh.circle-artifacts.com/0/.robot/report.html

and source code: https://github.com/helm/chartmuseum/tree/master/acceptance_tests

+1 to switching to Python or Go, but here is yet another option that I think would make the tool a little most robust long term and give us alot of flexibility.

**Comment by unguiculus** [comment]
@jdolitsky I like what you showed us in the call. I think Robot framework is a great thing and the reports are really nice. However, I'm afraid this is a massive thing with quite a learning curve which we should not impose on our users. We need something that's easy to use out of the box for our specific use case. Also, we really need verbose output and all the logging for charts CI. All logs, describe output etc. need to go into the log file because this is the only way users can see why tests fail.

**Comment by jdolitsky** [comment]
@unguiculus fair enough. carry on!

**Comment by scottrigby** [comment]
Also note if we want human/machine-readable testing scenarios in Go, here's go's Cucumber implementation: https://github.com/DATA-DOG/godog

And if we want HTML report output, there's https://github.com/gkushang/cucumber-html-reporter

**Comment by nvtkaszpir** [comment]
I believe doing this in python is better than rewriting it in go:
- IMHO python has much lower entry level than go
- we really do not need improved concurrency and performance
- with python we can use something ancient (that is stable!) as invoke (http://www.pyinvoke.org/, the successor of fabric)
- other people using python can tremendously benefit from utilizing other frameworks like tox, robot, cucumber...

Frankly speaking when I was looking through the code I just imagined how much easier would it be to do it in invoke or just with fabric. Bear in mind that I'm biased towards python, because I have not yet touched go that much ;)

**Comment by unguiculus** [comment]
@nvtkaszpir I did start in Python and am happy to have switched to Go.

**Comment by unguiculus** [comment]
Fixed in #35.
